redactevent.go

```diff
@@ -0,0 +1,131 @@
+		StateDefault  rawJSON `json:"state_default,omitempty"`
+		Ban           rawJSON `json:"ban,omitempty"`
+		Kick          rawJSON `json:"kick,omitempty"`
+		Redact        rawJSON `json:"redact,omitempty"`
+	}
+
+	// memberContent keeps the fields needed in a m.room.member event.
+	// Member events keep the membership.
+	// (In an ideal world they would keep the third_party_invite see matrix-org/synapse#1831)
+	type memberContent struct {
+		Membership rawJSON `json:"membership,omitempty"`
+	}
+
+	// aliasesContent keeps the fields needed in a m.room.aliases event.
+	// Alias events probably don't need to keep the aliases key, but we need to match synapse here.
+	type aliasesContent struct {
+		Aliases rawJSON `json:"aliases,omitempty"`
+	}
+
+	// historyVisibilityContent keeps the fields needed in a m.room.history_visibility event
+	// History visibility events need to keep the history_visibility key.
+	type historyVisibilityContent struct {
+		HistoryVisibility rawJSON `json:"history_visibility,omitempty"`
+	}
+
+	// allContent keeps the union of all the content fields needed across all the event types.
+	//All the content JSON keys we are keeping are distinct across the different event types.
+	type allContent struct {
+		createContent
+		joinRulesContent
+		powerLevelContent
+		memberContent
+		aliasesContent
+		historyVisibilityContent
+	}
+
+	// eventFields keeps the top level keys needed by all event types.
+	// (In an ideal world they would include the "redacts" key for m.room.redaction events, see matrix-org/synapse#1831)
+	type eventFields struct {
+		EventID        rawJSON    `json:"event_id,omitempty"`
+		Sender         rawJSON    `json:"sender,omitempty"`
+		RoomID         rawJSON    `json:"room_id,omitempty"`
+		Hashes         rawJSON    `json:"hashes,omitempty"`
+		Signatures     rawJSON    `json:"signatures,omitempty"`
+		Content        allContent `json:"content"`
+		Type           string     `json:"type"`
+		StateKey       rawJSON    `json:"state_key,omitempty"`
+		Depth          rawJSON    `json:"depth,omitempty"`
+		PrevEvents     rawJSON    `json:"prev_events,omitempty"`
+		PrevState      rawJSON    `json:"prev_state,omitempty"`
+		AuthEvents     rawJSON    `json:"auth_events,omitempty"`
+		Origin         rawJSON    `json:"origin,omitempty"`
+		OriginServerTS rawJSON    `json:"origin_server_ts,omitempty"`
+		Membership     rawJSON    `json:"membership,omitempty"`
+	}
+
+	var event eventFields
+	// Unmarshalling into a struct will discard any extra fields from the event.
+	if err := json.Unmarshal(eventJSON, &event); err != nil {
+		return nil, err
+	}
+	// Create a new, empty content object
+	var newContent allContent
+	// Copy the content fields that we should keep for the event type.
+	// By default we copy nothing leaving the content object empty.
+	switch event.Type {
+	case "m.room.create":
+		newContent.createContent = event.Content.createContent
+	case "m.room.member":
```

Comment:
Eh, it was one of the things I hated most about trying to read the synapse codebase fwiw.
It made it super hard to tell what was going on in the code because it meant that synapse was creating it's own names for everything rather than just using the strings from the matrix spec. But I agree that it would be nice if they were more resistant to being typo'd.